File: src/game/resources.js
import { LIMITS } from "./catalog.js";
import { bridgeGeometry } from "./geometry.js";
import { encounter } from "./identity.js";
import { noteEvidence } from "./story.js";
import { activity, postMessage } from "./community.js";
const types = { log: "wood", bone: "bones", ore: "ore", corpse: "corpses" };
export function releaseCargo(w, c) {
  if (c.carry > 0) {
    w.inventory[c.cargoKind || "wood"] += c.carry;
    c.carry = 0;
    c.cargoKind = null;
  }
}
export function deposit(w, type, x, y, stock, addObject) {
  const nearby = w.objects.find(
    (o) => o.type === type && Math.hypot(o.x - x, o.y - y) < 2,
  );
  if (nearby) {
    nearby.stock += stock;
    return nearby;
  }
  if (w.objects.length < LIMITS.objects)
    return addObject(w, type, x, y, { stock });
  // The recovery store is always available, even when every object slot is used.
  // A corpse is stored as a corpse, never silently converted to bones.
  const kind = types[type];
  if (kind) w.inventory[kind] = (w.inventory[kind] || 0) + stock;
  return null;
}
export function die(
  w,
  victims,
  cause,
  addObject,
  remember,
  actor = "caretaker",
) {
  const living = victims.filter((c) => w.creatures.some((o) => o.id === c.id));
  if (!living.length) return 0;
  for (const c of living) {
    releaseCargo(w, c);
    const sacrifice = cause === "bug" || cause === "swarm";
    deposit(
      w,
      sacrifice ? "bone" : "corpse",
      c.x,
      c.y,
      sacrifice ? 8 : 1,
      addObject,
    );
    for (const other of w.creatures)
      if (other.id !== c.id && Math.hypot(other.x - c.x, other.y - c.y) < 6)
        encounter(other, "loss", c.id, w.time);
    w.departed.push({ id: c.id, name: c.name, cause, tick: w.time, actor });
    remember(
      w,
      "loss",
      `${c.name} died (${cause}). ${sacrifice ? "Eight bones remain." : "Their remains are kept."}`,
      c.id,
    );
    noteEvidence(w, "deaths", `${c.name} died from ${cause}.`, 1, c.id);
    const dimension = sacrifice
      ? "sacrificed"
      : cause === "hammer"
        ? "hammer"
        : cause === "meteor"
          ? "meteorDeaths"
        : cause === "pollution"
          ? "pollutionDeaths"
          : "neglect";
    noteEvidence(w, dimension, `${c.name}: ${cause}.`, 1, c.id);
  }
  const ids = new Set(living.map((c) => c.id));
  w.creatures = w.creatures.filter((c) => !ids.has(c.id));
  w.departed = w.departed.slice(-32);
  w.population = w.creatures.length + w.cohort + w.orbital.population;
  w.commandRevision++;
  return living.length;
}
export function deliver(w, c, o, remember) {
  const g = bridgeGeometry(o),
    kind = c.cargoKind || "wood";
  if (g.complete) {
    releaseCargo(w, c);
    return;
  }
  const remaining = g.required - g.delivered.wood - g.delivered.bones,
    amount = Math.min(c.carry, remaining);
  g.delivered[kind] = (g.delivered[kind] || 0) + amount;
  c.carry -= amount;
  noteEvidence(
    w,
    "deliveries",
    `${c.name} delivered ${amount} ${kind}.`,
    1,
    c.id,
  );
  noteEvidence(
    w,
    kind === "bones" ? "bonesDelivered" : "woodDelivered",
    `${amount} ${kind} became part of the bridge.`,
    amount,
    c.id,
  );
  o.stock = g.delivered.wood + g.delivered.bones;
  activity(w, "bridge", `${c.name} delivered ${amount} ${kind}. Bridge: ${o.stock}/${g.required}.`);
  if (o.stock >= g.required) {
    g.complete = true;
    w.progress.bridge = true;
    w.stage = Math.max(w.stage, 2);
    w.navRevision++;
    w.commandRevision++;
    remember(
      w,
      "bridge",
      "The bridge is complete. Both banks are connected.",
      o.id,
    );
    postMessage(w, { key:`bridge-complete:${o.id}`, title:"The other bank",
      text:"We finished the crossing. The ore deposits and mountain are within reach now. Our world has room to grow." });
  }
  o.bridge = g;
  if (c.carry) releaseCargo(w, c);
  else c.cargoKind = null;
}
export function pollutionAt(w, p) {
  return w.pollution.reduce(
    (sum, z) =>
      sum +
      z.amount * Math.max(0, 1 - Math.hypot(p.x - z.x, p.y - z.y) / z.radius),
    0,
  );
}
export function pollute(w, o, amount) {
  let z = w.pollution.find((z) => z.source === o.id);
  if (!z) {
    z = { source: o.id, x: o.x, y: o.y, radius: 5, amount: 0 };
    w.pollution.push(z);
  }
  z.amount = Math.min(250, z.amount + amount);
  w.pollution = w.pollution.slice(-64);
  w.progress.pollution = Math.min(
    1000,
    w.pollution.reduce((s, z) => s + z.amount, 0),
  );
}
export function cleanPollution(w, p, remember) {
  let removed = 0;
  for (const z of w.pollution)
    if (Math.hypot(z.x - p.x, z.y - p.y) < z.radius + 3) {
      const n = Math.min(40, z.amount);
      z.amount -= n;
      removed += n;
    }
  w.pollution = w.pollution.filter((z) => z.amount > 0.1);
  w.progress.pollution = w.pollution.reduce((s, z) => s + z.amount, 0);
  for (const c of w.creatures)
    if (Math.hypot(c.x - p.x, c.y - p.y) < 4) {
      c.clean = Math.max(c.clean, 90);
      c.sickness = Math.max(0, c.sickness - 30);
    }
  if (removed) {
    noteEvidence(
      w,
      "cleaned",
      `Cleaned ${Math.round(removed)} pollution near (${Math.round(p.x)}, ${Math.round(p.y)}).`,
      removed,
    );
    remember(
      w,
      "cleanup",
      "The air clears around this part of the settlement.",
    );
  }
}

// Worker maintenance removes only actual pollution at the serviced factory.
// No free need restoration, material creation, or per-resident history entries.
export function maintainFactory(w, factory, amount) {
  const zone = w.pollution.find((z) => z.source === factory.id);
  if (!zone) return 0;
  const removed = Math.min(Math.max(0, amount), zone.amount);
  zone.amount -= removed;
  if (zone.amount <= 0) w.pollution = w.pollution.filter((z) => z !== zone);
  w.progress.pollution = w.pollution.reduce((sum, z) => sum + z.amount, 0);
  return removed;
}

File: engine/src/resources.rs
use crate::{
    Engine,
    geometry::{Object, bridge},
    identity::encounter,
    value::*,
};
use serde_json::{Value, json};

pub(crate) fn bridge_state(o: &Value) -> Value {
    let g = bridge(&Object::read(o));
    json!({"a":g.a,"b":g.b,"width":g.width,"elevation":0.18,
      "required":o["bridge"]["required"].as_f64().unwrap_or(24.),
      "delivered":o["bridge"].get("delivered").filter(|x|!x.is_null()).cloned().unwrap_or(json!({"wood":num(o,"stock"),"bones":0})),"complete":g.complete})
}
pub(crate) fn accepts_project(o: &Value, kind: &str) -> bool {
    text(o, "type") == "sculpture" && num(o, "stock") < 12. && kind == "wood"
}
impl Engine {
    pub(crate) fn release_cargo(&mut self, c: &mut Value) {
        if num(c, "carry") > 0. {
            let kind = if text(c, "cargoKind").is_empty() {
                "wood"
            } else {
                text(c, "cargoKind")
            }
            .to_string();
            increment(&mut self.world["inventory"], &kind, num(c, "carry"));
            c["carry"] = json!(0);
            c["cargoKind"] = Value::Null;
        }
    }
    pub(crate) fn deposit(&mut self, kind: &str, p: Point, stock: f64) -> Option<Value> {
        if let Some(o) = self.world["objects"].as_array_mut().and_then(|a| {
            a.iter_mut()
                .find(|o| text(o, "type") == kind && Point::read(o).distance(p) < 2.)
        }) {
            increment(o, "stock", stock);
            return Some(o.clone());
        }
        if list(&self.world, "objects").len() < 2048 {
            return self.add_object(kind, p.x, p.y, json!({"stock":stock}));
        }
        let material = match kind {
            "log" => "wood",
            "bone" => "bones",
            "ore" => "ore",
            "corpse" => "corpses",
            _ => return None,
        };
        increment(&mut self.world["inventory"], material, stock);
        None
    }
    pub(crate) fn die(&mut self, victims: &[Value], cause: &str, actor: &str) -> usize {
        let living: Vec<_> = victims
            .iter()
            .filter(|c| {
                list(&self.world, "creatures")
                    .iter()
                    .any(|o| same_id(&o["id"], &c["id"]))
            })
            .cloned()
            .collect();
        let time = num(&self.world, "time");
        for mut c in living.clone() {
            self.density_index.borrow_mut().remember_position(&c);
            self.release_cargo(&mut c);
            let sacrifice = matches!(cause, "bug" | "swarm");
            let p = Point::read(&c);
            self.deposit(
                if sacrifice { "bone" } else { "corpse" },
                p,
                if sacrifice { 8. } else { 1. },
            );
            for other in self.world["creatures"].as_array_mut().into_iter().flatten() {
                if !same_id(&other["id"], &c["id"]) && Point::read(other).distance(p) < 6. {
                    encounter(other, "loss", Some(text(&c, "id")), time);
                }
            }
            self.world["departed"].as_array_mut().unwrap().push(
                json!({"id":c["id"],"name":c["name"],"cause":cause,"tick":time,"actor":actor}),
            );
            self.remember(
                "loss",
                &format!(
                    "{} died ({}). {}",
                    text(&c, "name"),
                    cause,
                    if sacrifice {
                        "Eight bones remain."
                    } else {
                        "Their remains are kept."
                    }
                ),
                Some(text(&c, "id")),
            );
            self.note_evidence(
                "deaths",
                &format!("{} died from {}.", text(&c, "name"), cause),
                1.,
                Some(text(&c, "id")),
            );
            let dimension = if sacrifice {
                "sacrificed"
            } else {
                match cause {
                    "hammer" => "hammer",
                    "meteor" => "meteorDeaths",
                    "pollution" => "pollutionDeaths",
                    _ => "neglect",
                }
            };
            self.note_evidence(
                dimension,
                &format!("{}: {}.", text(&c, "name"), cause),
                1.,
                Some(text(&c, "id")),
            );
        }
        if living.is_empty() {
            return 0;
        }
        self.world["creatures"]
            .as_array_mut()
            .unwrap()
            .retain(|c| !living.iter().any(|o| same_id(&o["id"], &c["id"])));
        let departed = self.world["departed"].as_array_mut().unwrap();
        if departed.len() > 32 {
            departed.drain(..departed.len() - 32);
        }
        let population = list(&self.world, "creatures").len() as f64
            + num(&self.world, "cohort")
            + num(&self.world["orbital"], "population");
        self.world["population"] = json!(population);
        increment(&mut self.world, "commandRevision", 1.);
        living.len()
    }
    pub(crate) fn deliver(&mut self, c: &mut Value, o: &mut Value) {
        let mut g = bridge_state(o);
        let kind = if text(c, "cargoKind").is_empty() {
            "wood"
        } else {
            text(c, "cargoKind")
        }
        .to_string();
        if flag(&g, "complete") {
            self.release_cargo(c);
            return;
        }
        let remaining =
            num(&g, "required") - num(&g["delivered"], "wood") - num(&g["delivered"], "bones");
        let amount = num(c, "carry").min(remaining);
        increment(&mut g["delivered"], &kind, amount);
        increment(c, "carry", -amount);
        self.note_evidence(
            "deliveries",
            &format!("{} delivered {} {}.", text(c, "name"), amount, kind),
            1.,
            Some(text(c, "id")),
        );
        self.note_evidence(
            if kind == "bones" {
                "bonesDelivered"
            } else {
                "woodDelivered"
            },
            &format!("{} {} became part of the bridge.", amount, kind),
            amount,
            Some(text(c, "id")),
        );
        o["stock"] = json!(num(&g["delivered"], "wood") + num(&g["delivered"], "bones"));
        self.activity(
            "bridge",
            &format!(
                "{} delivered {} {}. Bridge: {}/{}.",
                text(c, "name"),
                amount,
                kind,
                num(o, "stock"),
                num(&g, "required")
            ),
            "Instincts",
            "",
        );
        if num(o, "stock") >= num(&g, "required") {
            g["complete"] = json!(true);
            self.world["progress"]["bridge"] = json!(true);
            self.world["stage"] = json!(num(&self.world, "stage").max(2.));
            increment(&mut self.world, "navRevision", 1.);
            increment(&mut self.world, "commandRevision", 1.);
            self.remember(
                "bridge",
                "The bridge is complete. Both banks are connected.",
                Some(text(o, "id")),
            );
            self.post_message(json!({"key":format!("bridge-complete:{}",text(o,"id")),"title":"The other bank","text":"We finished the crossing. The ore deposits and mountain are within reach now. Our world has room to grow."}));
        }
        o["bridge"] = g;
        if num(c, "carry") > 0. {
            self.release_cargo(c);
        } else {
            c["cargoKind"] = Value::Null;
        }
    }
    pub(crate) fn supply_project(&mut self, c: &mut Value, o: &mut Value) -> bool {
        if !accepts_project(o, text(c, "cargoKind")) {
            return false;
        }
        let used = num(c, "carry").min(12. - num(o, "stock"));
        increment(o, "stock", used);
        increment(c, "carry", -used);
        if num(c, "carry") == 0. {
            c["cargoKind"] = Value::Null;
        }
        if num(o, "stock") >= 12. {
            let message = "The sculpture keeps the shape of our first question.";
            self.remember("project", message, Some(text(o, "id")));
            self.note_evidence("discovered", message, 1., Some(text(o, "id")));
            o["discovered"] = json!(true);
        }
        true
    }
    pub(crate) fn pollution_at(&self, p: Point) -> f64 {
        list(&self.world, "pollution")
            .iter()
            .map(|z| {
                num(z, "amount") * (1. - Point::read(z).distance(p) / num(z, "radius")).max(0.)
            })
            .sum()
    }
    pub(crate) fn pollute(&mut self, o: &Value, amount: f64) {
        let pollution = self.world["pollution"].as_array_mut().unwrap();
        let index = if let Some(i) = pollution
            .iter()
            .position(|z| same_id(&z["source"], &o["id"]))
        {
            i
        } else {
            pollution.push(json!({"source":o["id"],"x":o["x"],"y":o["y"],"radius":5,"amount":0}));
            pollution.len() - 1
        };
        let z = &mut pollution[index];
        z["amount"] = json!((num(z, "amount") + amount).min(250.));
        if pollution.len() > 64 {
            pollution.drain(..pollution.len() - 64);
        }
        self.world["progress"]["pollution"] = json!(
            list(&self.world, "pollution")
                .iter()
                .map(|z| num(z, "amount"))
                .sum::<f64>()
                .min(1000.)
        );
    }
    pub(crate) fn clean_pollution(&mut self, p: Point) {
        let mut removed = 0.;
        for z in self.world["pollution"].as_array_mut().into_iter().flatten() {
            if Point::read(z).distance(p) < num(z, "radius") + 3. {
                let n = num(z, "amount").min(40.);
                increment(z, "amount", -n);
                removed += n;
            }
        }
        self.world["pollution"]
            .as_array_mut()
            .unwrap()
            .retain(|z| num(z, "amount") > 0.1);
        self.world["progress"]["pollution"] = json!(
            list(&self.world, "pollution")
                .iter()
                .map(|z| num(z, "amount"))
                .sum::<f64>()
        );
        for c in self.world["creatures"].as_array_mut().into_iter().flatten() {
            if Point::read(c).distance(p) < 4. {
                c["clean"] = json!(num(c, "clean").max(90.));
                c["sickness"] = json!((num(c, "sickness") - 30.).max(0.));
            }
        }
        if removed > 0. {
            self.note_evidence(
                "cleaned",
                &format!(
                    "Cleaned {} pollution near ({}, {}).",
                    crate::simulation::js_round(removed),
                    crate::simulation::js_round(p.x),
                    crate::simulation::js_round(p.y)
                ),
                removed,
                None,
            );
            self.remember(
                "cleanup",
                "The air clears around this part of the settlement.",
                None,
            );
        }
    }
    pub(crate) fn maintain_factory(&mut self, o: &Value, amount: f64) -> f64 {
        let zones = self.world["pollution"].as_array_mut().unwrap();
        let Some(z) = zones.iter_mut().find(|z| same_id(&z["source"], &o["id"])) else {
            return 0.;
        };
        let removed = amount.max(0.).min(num(z, "amount"));
        increment(z, "amount", -removed);
        zones.retain(|z| num(z, "amount") > 0.);
        self.world["progress"]["pollution"] = json!(
            list(&self.world, "pollution")
                .iter()
                .map(|z| num(z, "amount"))
                .sum::<f64>()
        );
        removed
    }
}
